source/more_sampling_tools.Rmd

```diff
@@ -209,13 +211,513 @@ message(n_golds / (n_golds + n_silvers))
+
+```{python}
+# First question, giving first element of result.
+ten_first_gold[0] & ten_second_gold[0]
+```
+
+```{r}
+# First question, giving first element of result.
+ten_first_gold[1] & ten_second_gold[1]
+```
+
+The second question is:
+
+```{python}
+# Second question, giving second element of result.
+ten_first_gold[1] & ten_second_gold[1]
+```
+
+```{python}
+# Second question, giving second element of result.
+ten_first_gold[2] & ten_second_gold[2]
+```
+
+and so on.  We have ten elements on *each side*, and 10 answers, giving {{< var
+an_array >}} (`ten_both`) of 10 elements.  Each element in `ten_both` is the
+answer to the `&` question for the elements at the corresponding positions in
+`ten_first_gold` and `ten_second_gold`.
+
+We could also create the Boolean {{< var array >}}s and do the `&` operation
+all in one step, like this:
+
+```{python}
+ten_both = (ten_first_trunks == 'Gold') & (ten_second_trunks == 'Gold')
+ten_both
+```
+
+<!---
+Another illustration here, of form (matching the text):
+
+True          True          True
+False         False         True
+False         False         True
+True          True          False
+False    =    False    &    True
+...
+-->
+
+
+::: python
+
+:::{.callout-note}
+## Parentheses, arrays and comparisons
+
+Notice the round brackets (parentheses) around `(ten_first_trunks == 'Gold')`
+and `(ten_second_trunks == 'Gold')`.  In fact, you need these to make Python do
+what we want here.   Remember *operator precedence*; the multiply operator `*`
+has *higher precedence* than the operator `+`, so `3 + 5 * 2` is equal to 3 +
+(5 * 2) = 13.  If we want to do addition before multiplication, we use
+parentheses to tell Python the order it should use: `(3 + 5) * 2` = 16.
+
+The same applies for the two operators `==` and `&` here.   In fact `&` has a
+higher precedence than `==` so Python will read the expression without
+parentheses — `ten_first_trunks == 'Gold' & ten_second_trunks == 'Gold'` to
+mean (because of operator precedence) `ten_first_trunks == ('Gold' &
+ten_second_trunks) == 'Gold'`; it does not know what to do with `'Gold' &
+ten_second_trunks` and generates an error:
+
+```{python tags=c("raises-exception"), error=TRUE}
+# The statement without parentheses generates an error, because of the
+# "Gold' & ten_second_trunks operation, from operator precedence.
+ten_first_trunks == 'Gold' & ten_second_trunks == 'Gold'
+```
+
+This is the same error you get for:
+
+```{python tags=c("raises-exception"), error=TRUE}
+# The expression that causes the error.
+'Gold' & ten_second_trunks
+```
+
+The point to take away is, that when you are using `&` to combine Boolean
+arrays in Python, remember operator precedence, and, when in doubt, put
```

Comment:
Isn't it good practice to use parentheses anyway?

Reply:
OK - I'll add something there about good practice.